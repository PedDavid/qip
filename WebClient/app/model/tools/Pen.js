import { Figure, FigureStyle } from './../Figure'
import {SimplePoint} from './../SimplePoint'
import {PointStyle} from './../Point'
import Tool from './Tool'
import SettingsConfig from './../../util/SettingsConfig.js'

export default class Pen implements Tool {
  constructor (grid, color, width) {
    this.type = 'pen'
    this.color = color
    this.width = width
    this.fontText = width

    this.grid = grid
    this.currentFigure = null
  }

  // passar para aqui as funções getMousePos
  onPress (event, scale, updateCanvasSizeFunc, settings) {
    if (settings[SettingsConfig.fingerMoveSettingIdx] && event.pointerType === 'touch') {
      return
    }
    const x = event.offsetX
    const y = event.offsetY

    const figStyle = new FigureStyle(this.color, scale)
    // Create a new Figure
    this.currentFigure = new Figure(figStyle)

    const press = event.pressure * this.width
    const pointStyle = new PointStyle(press)
    this.currentFigure.addPoint(new SimplePoint(x, y, pointStyle))

    const canvasContext = event.target.getContext('2d')
    canvasContext.beginPath()
    canvasContext.arc(x, y, press / 4, 0, 2 * Math.PI, false)
    canvasContext.fillStyle = this.color
    canvasContext.fill()
    canvasContext.lineWidth = press / 2
    canvasContext.strokeStyle = this.color
    canvasContext.stroke()

    updateCanvasSizeFunc(x, y)
  }

  onSwipe (event, scale, updateCanvasSizeFunc, settings) {
    if (event.buttons > 0 && !(settings[SettingsConfig.fingerMoveSettingIdx] && event.pointerType === 'touch')) {
      const x = event.offsetX
      const y = event.offsetY

      if (this.currentFigure === null) {
        return
      }

      const press = event.pressure * this.width
      const pointStyle = new PointStyle(press)
      this.currentFigure.points.length === 0 &&
        this.currentFigure.addPoint(new SimplePoint(x, y, pointStyle)) // this is necessary to avoid a bug when user draws out of the screen
      const last = this.currentFigure.points[this.currentFigure.points.length - 1]
      // Ignore if swiped to last coordinates
      if (last.x === x && last.y === y) {
        return
      }

      this.currentFigure.addPoint(new SimplePoint(x, y, pointStyle))

      const canvasContext = event.target.getContext('2d')
      canvasContext.beginPath() // também tem de estar aqui para dar para fazer pressure sensitive

      canvasContext.moveTo(last.x, last.y)
      canvasContext.lineTo(x, y)

      canvasContext.lineWidth = press
      canvasContext.lineJoin = canvasContext.lineCap = 'round'
      canvasContext.stroke()

      updateCanvasSizeFunc(x, y)
    }
  }

  onPressUp (event, persist) {
    if (this.currentFigure === null) { // it is necessary to avoid some minor bugs
      return
    }
    // todo: passar persistencia para o onOut
    this.grid.addFigure(this.currentFigure, true)

    persist.sendPenAction(this.currentFigure, this.grid.getCurrentFigureId())

    // reset current figure
    this.currentFigure = null
  }

  onOut (event, persist) {
    if (this.currentFigure === null || this.currentFigure.points.length <= 0) {
      return
    }
    this.grid.addFigure(this.currentFigure, true)

    persist.sendPenAction(this.currentFigure, this.grid.getCurrentFigureId())

    // fazer reset à figura para que não continue a desenhar se o utilizador sair da área do canvas
    // Desta forma, se o utilizador voltar à área do canvas com o ponteiro premido, irá desenhar uma nova figura
    const figStyle = new FigureStyle(this.color)
    this.currentFigure = new Figure(figStyle)
  }

  onContextMenu (event) {}

  equals (pen) {
    return pen instanceof Pen && this.width === pen.width && this.color === pen.color
  }
}
