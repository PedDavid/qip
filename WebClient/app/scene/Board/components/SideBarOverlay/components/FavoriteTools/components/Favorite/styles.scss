.fav {
    width: 40px;
    height: 40px;
    display: inline-block;
}
.fav:hover {
  transform: rotate(360deg) translate( 0px, 0px);
  transition: 0.5s;
}

// for some reason, when favorite is being rotated, icon and font are moving. Therefore, this compensates child elements 
.fav:hover > .iconStyle {
  left: 7px;
}
.fav:hover > .fontStyle {
  bottom: 0px;
}

.favMenu {
  z-index: -1;
  width: 130px;
  height: 40px;
  float: left;
  position: absolute;
  transition: 0.5s;
  text-align: right;
}

.fontStyle {
  position: absolute;
  bottom: 5px;
}

.iconStyle {
  position: absolute;
  top: 10px;
  left: 25px;
}
