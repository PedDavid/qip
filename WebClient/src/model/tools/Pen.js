import { PointStyle } from './../Point'
import { Figure, FigureStyle } from './../Figure'

export function Pen (grid, canvasContext, onUpdateCanvasSize) {
  this.currentFigure = null
  // passar para aqui as funções getMousePos
  // passar como param updateCanvasSize??
  this.onDown = function (currentItem, coord, event, scale) {
    const figStyle = new FigureStyle(currentItem.pen.name, scale)
    // Create a new Figure
    this.currentFigure = new Figure(grid.getNewFigureIdx(), null, figStyle)

    let press = event.pressure * currentItem.pen.width
    const pointStyle = new PointStyle(press)
    // Get the point
    const point = grid.getOrCreatePoint(coord.x, coord.y)
    // Add point to the Figure
    this.currentFigure.addPoint(point)
    // Add Figure to the point
    point.addFigure(this.currentFigure.id, pointStyle)

    canvasContext.beginPath()
    canvasContext.arc(coord.x, coord.y, press / 4, 0, 2 * Math.PI, false)
    canvasContext.fillStyle = currentItem.pen.name
    canvasContext.fill()
    canvasContext.lineWidth = press / 2
    canvasContext.strokeStyle = currentItem.pen.name
    canvasContext.stroke()
    onUpdateCanvasSize(coord.x, coord.y)
  }

  this.onMove = function (currentItem, coord, event, scale) {
    if (event.buttons > 0) {
      if (this.currentFigure == null) {
        return
      }

      let press = event.pressure * currentItem.pen.width
      const pointStyle = new PointStyle(press)

      canvasContext.beginPath() // também tem de estar aqui para dar para fazer pressure sensitive

      const point = grid.getOrCreatePoint(coord.x, coord.y)

      // esta situação pode acontecer quando o utilizador sai fora dos limites do canvas
      // caso aconteça, acrescentar o primeiro ponto
      if (this.currentFigure.points.length === 0) {
        point.addFigure(this.currentFigure.id, pointStyle)
        this.currentFigure.addPoint(point)
        return
      }

      const last = this.currentFigure.points[this.currentFigure.points.length - 1]

      if (last.x === point.x && last.y === point.y) {   // verificar se o evento ocorreu nas mesmas coordenadas que anteriormente
        return                                        // nesse caso ignorá-lo
      }

      point.addFigure(this.currentFigure.id, pointStyle)
      this.currentFigure.addPoint(point)

      grid.updateMaxLinePart(last, point, this.currentFigure, pointStyle) // atualizar a variável que guarda o valor máximo entre pontos intermédios de retas
      const width = event.pressure * currentItem.pen.width

      canvasContext.moveTo(last.x, last.y)
      canvasContext.lineTo(point.x, point.y)

      canvasContext.lineWidth = width
      canvasContext.lineJoin = canvasContext.lineCap = 'round'
      canvasContext.stroke()

      onUpdateCanvasSize(coord.x, coord.y)
    }
  }

  this.onUp = function () {
    grid.addFigure(this.currentFigure)
    this.currentFigure = null
  }

  this.onOut = function (currentItem) {
    if (this.currentFigure == null) {
      return
    }
    grid.addFigure(this.currentFigure)

    // fazer reset à figura para que não continue a desenhar se o utilizador sair da área do canvas
    // Desta forma, se o utilizador voltar à área do canvas com o ponteiro premido, irá desenhar uma nova figura
    const figStyle = new FigureStyle(currentItem.pen.name)
    this.currentFigure = new Figure(grid.getNewFigureIdx(), null, figStyle)
  }
}
