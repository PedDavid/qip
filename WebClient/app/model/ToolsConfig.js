import Tool from './tools/Tool'
import Pen from './tools/Pen'
import Eraser from './tools/Eraser'

// todo (simaovii) tirar esta classe do model. por em utils

// this class has the purpose to configure all tools and simplify all the management of the tools.
// DefaultTools can be accessed two ways:
// by calling ToolsConfig.defaultTools -> Tool[]
// by calling ToolsConfig[pen|eraser|move] -> ConfigTool[] with pens
export default class ToolsConfig {
  constructor (defaultToolsRaw) {
    this.defaultTools = defaultToolsRaw
    this._mapTools(defaultToolsRaw)
  }

  // !!! at the moment, this is not working
  getDefaultToolOf = function (toolType: Tool) {
    const idx = this.defaultTools
      .findIndex(defTool => {
        if (defTool.toolType !== undefined) {
          return defTool.toolType.name === toolType
        }
        return false
      })
    return this.defaultTools[idx]
  }

  // private method to map all the defaultTools to a type so it can be possible
  // to do calls like: ToolsConfig[pen] ...
  _mapTools = function (defaultTools) {
    defaultTools.forEach(defTool => {
      this[defTool.type] = new DefaultTool(defTool.type, defTool.icon, defTool.content, defTool.lastValue, defTool.toolType)
    })
  }

  // method to facilitate the change of tool. This way it can be called only with one
  // parameter. It will automatically set the previous tool of each tool type
  updatePrevTool = function (prevTool) {
    if (prevTool instanceof Pen) {
      this.pen.lastValue = prevTool
      this.width.lastValue = prevTool // mantain this here to be easier to update width
    } else if (prevTool instanceof Eraser) {
      this.eraser.lastValue = prevTool
    }
  }
}

// this class represents a DefaultTool.
// With this class, when writting code, it will be simplier to get tools' properties
class DefaultTool {
  constructor (type, icon, content, lastValue, toolType: Tool) {
    this.type = type
    this.icon = icon
    this.content = content
    this.lastValue = lastValue
    this.toolType = toolType
  }
}
