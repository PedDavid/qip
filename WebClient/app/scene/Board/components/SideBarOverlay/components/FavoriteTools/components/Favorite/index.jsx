import React from 'react'
import styles from './styles.scss'

import { Button, Icon } from 'semantic-ui-react'

export default class Favorite extends React.Component {
  state = {
    openFavMenu: false,
    animateClass: styles.fav
  }
  componentDidMount () {
    this.refs.btnDiv.addEventListener('webkitAnimationEnd', this.endAnimate)
  }
  animate = () => {
    this.setState({animateClass: styles.animateFav})
  }
  endAnimate = () => {
    this.setState({animateClass: ''})
  }
  toggleFavMenu = (evt) => {
    evt.preventDefault()
    this.setState(prevState => ({openFavMenu: !prevState.openFavMenu}))
  }
  render () {
    const favMenuWidth = this.state.openFavMenu ? 'visible' : 'hidden'
    const fav = this.props.fav

    const iconName = this.props.toolsConfig.getDefaultToolOf(fav.type).icon
    const color = fav.color !== undefined ? fav.color : 'black'

    let style = null
    if (this.props.currTool.equals(fav)) {
      style = {
        backgroundColor: 'gray'
      }
    }
    return (
      <div ref='btnDiv' className={this.state.animateClass} onContextMenu={this.toggleFavMenu}>
        <Button circular className={styles.fav} style={style} onClick={() => this.props.changeCurrentTool(fav)}>
          {/* it could be color:red instead of style:{color:'red'}} but the first one does not support rgba */}
          <Icon className={styles.iconStyle + ' large'} name={iconName} style={{color: color}} />
          <font className={styles.fontStyle} size='1'> {fav.width} </font>
        </Button>
        <Button onClick={() => this.props.removeFavorite(fav)}
          icon={{name: 'trash', style: {color: 'red'}}} content=' Remove'
          style={{marginLeft: '-25px', visibility: favMenuWidth}} className={styles.favMenu} />
      </div>
    )
  }
}
