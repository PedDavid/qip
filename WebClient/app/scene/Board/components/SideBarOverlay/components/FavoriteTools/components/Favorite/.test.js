import React from 'react'
import { shallow, mount } from 'enzyme'

import Favorite from '.'

import { Button } from 'semantic-ui-react'

import Pen from './../../../../../../../../model/tools/Pen'

import ToolsConfig from './../../../../../../../../model/ToolsConfig'
import defaultToolsConfig from './../../../../../../../../public/configFiles/defaultTools'


describe('Component: Favorite', () => {
  const tool = new Pen(null, 'red', 5)
  const toolsConfig = new ToolsConfig(defaultToolsConfig)
  const props = {
    fav: tool,
    currTool: tool,
    toolsConfig
  }
  const wrapper = shallow(<Favorite {...props} />)
  it('Is not null', () => {
    expect(wrapper).not.toBeNull()
  })
  it('Is instance of Favorite', () => {
    expect(wrapper.instance()).toBeInstanceOf(Favorite)
  })
  it('Type is <div>', () => {
    expect(wrapper.type()).toBe('div')
  })
  it('Props are delivered', () => {
    expect(wrapper.instance().props).toMatchObject(props)
  })
  it('Renders as expected', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('Has 2 buttons', () => {
    expect(wrapper.find(Button)).toHaveLength(2)
  })
  it.skip('Check type of tool property', () => {})
  it('State starts with property openFavMenu as false', () => {
    const wrapper = shallow(<Favorite {...props} />)
    expect(wrapper.state().openFavMenu).toBe(false)
  })
  it('Listens to onContextMenu', () => {
    const wrapper = shallow(<Favorite {...props} />)
    wrapper.simulate('contextMenu', {preventDefault: () => {}})
    expect(wrapper.state().openFavMenu).toBe(true)
  })
  it('Listens to buttons onClicks, has changeCurrentTool and removeFavorite listeners', () => {
    expect.assertions(2)
    const mProps = {
      ...props,
      changeCurrentTool: fav => {
        expect(fav).toBe(props.fav)
      },
      removeFavorite: fav => {
        expect(fav).toBe(props.fav)
      }
    }
    const wrapper = shallow(<Favorite {...mProps} />)
    wrapper.find(Button).forEach(buttonWrapper => buttonWrapper.simulate('click'))
  })
})