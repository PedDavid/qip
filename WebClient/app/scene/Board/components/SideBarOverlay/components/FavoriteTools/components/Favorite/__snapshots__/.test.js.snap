// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Favorite Renders as expected 1`] = `"<div class=\\"fav\\"><button style=\\"background-color:gray;\\" class=\\"ui circular button fav\\"><i style=\\"color:red;\\" aria-hidden=\\"true\\" class=\\"pencil icon iconStyle large\\"></i><font class=\\"fontStyle\\" size=\\"1\\"> 5 </font></button><button style=\\"margin-left:-25px;visibility:hidden;\\" class=\\"ui button favMenu\\"><i style=\\"color:red;\\" aria-hidden=\\"true\\" class=\\"trash icon\\"></i>  Remove</button></div>"`;
